Not compatible to OXID eShop 6.2

The module extends e.g. 
`OxidEsales\EshopCommunity\Core\Module\ModuleInstaller`. In this class, the PhpDoc states: "@internal Do not make a module extension for this class." With OXID eShop v6.2.0-beta.1 (including https://github.com/OXID-eSales/oxideshop_ce/tree/v6.4.0), e.g. the method `OxidEsales\EshopCommunity\Core\Module\ModuleInstaller::validateMetadataExtendSection()` was removed. Therefore, the module does not work anymore. Maybe there are also other issues. 

As the whole module installation and activation changed with OXID eShop v6.2.0-beta.1, the strategy how to proceed with this module must be reviewed.

I've now started to work on this, here are the test results for OXID 6.2
```
running php version 7.1.33

============
Adding unit tests from /var/www/OXID/source/modules/oxcom/moduleinternals/Tests//Integration/Controller/*Test\.php
Adding unit tests from /var/www/OXID/source/modules/oxcom/moduleinternals/Tests//Integration/Core/*Test\.php
PHPUnit 6.5.14 by Sebastian Bergmann and contributors.

..E....E.EEEEEEE                                                  16 / 16 (100%)

Time: 2.37 seconds, Memory: 24.00MB

There were 9 errors:

1) OxidCommunity\ModuleInternals\Tests\Integration\Controller\NavigationControllerTest::testNavigationController
OxidEsales\Eshop\Core\Exception\SystemComponentException: Function 'getClassProviderStorage' does not exist or is not accessible! (OxidCommunity\ModuleInternals\Core\ModuleStateFixer)


/var/www/OXID/vendor/oxid-esales/oxideshop-ce/source/Core/Base.php:81
/__w/oxid-module-internals/oxid-module-internals/Core/ModuleStateFixer.php:492
/__w/oxid-module-internals/oxid-module-internals/Core/ModuleStateFixer.php:499
/__w/oxid-module-internals/oxid-module-internals/Core/ModuleStateFixer.php:127
/__w/oxid-module-internals/oxid-module-internals/Controller/Admin/NavigationController.php:44
/var/www/OXID/source/modules/oxcom/moduleinternals/Tests/Integration/Controller/NavigationControllerTest.php:16
/var/www/OXID/vendor/oxid-esales/testing-library/library/UnitTestCase.php:152

2) OxidCommunity\ModuleInternals\Tests\Integration\Controller\UtilsControllerTest::testGetModuleInstaller
TypeError: Argument 1 passed to OxidEsales\EshopCommunity\Internal\Framework\Module\Configuration\Bridge\ModuleConfigurationDaoBridge::get() must be of the type string, null given, called in /var/www/OXID/vendor/oxid-esales/oxideshop-ce/source/Core/Module/Module.php on line 108

/var/www/OXID/vendor/oxid-esales/oxideshop-ce/source/Internal/Framework/Module/Configuration/Bridge/ModuleConfigurationDaoBridge.php:50
/var/www/OXID/vendor/oxid-esales/oxideshop-ce/source/Core/Module/Module.php:108
/__w/oxid-module-internals/oxid-module-internals/Core/InternalModule.php:43
/__w/oxid-module-internals/oxid-module-internals/Controller/Admin/UtilsController.php:57
/__w/oxid-module-internals/oxid-module-internals/Controller/Admin/UtilsController.php:71
/__w/oxid-module-internals/oxid-module-internals/Controller/Admin/UtilsController.php:85
/var/www/OXID/source/modules/oxcom/moduleinternals/Tests/Integration/Controller/UtilsControllerTest.php:71
/var/www/OXID/vendor/oxid-esales/testing-library/library/UnitTestCase.php:152

3) OxidCommunity\ModuleInternals\Tests\Integration\Controller\UtilsControllerTest::testActivateModule
TypeError: Argument 1 passed to OxidEsales\EshopCommunity\Internal\Framework\Module\Configuration\Bridge\ModuleConfigurationDaoBridge::get() must be of the type string, null given, called in /var/www/OXID/vendor/oxid-esales/oxideshop-ce/source/Core/Module/Module.php on line 108

/var/www/OXID/vendor/oxid-esales/oxideshop-ce/source/Internal/Framework/Module/Configuration/Bridge/ModuleConfigurationDaoBridge.php:50
/var/www/OXID/vendor/oxid-esales/oxideshop-ce/source/Core/Module/Module.php:108
/__w/oxid-module-internals/oxid-module-internals/Core/InternalModule.php:43
/__w/oxid-module-internals/oxid-module-internals/Controller/Admin/UtilsController.php:57
/__w/oxid-module-internals/oxid-module-internals/Controller/Admin/UtilsController.php:118
/var/www/OXID/source/modules/oxcom/moduleinternals/Tests/Integration/Controller/UtilsControllerTest.php:131
/var/www/OXID/vendor/oxid-esales/testing-library/library/UnitTestCase.php:152

4) OxidCommunity\ModuleInternals\Tests\Integration\Controller\UtilsControllerTest::testDeactivateModule
TypeError: Argument 1 passed to OxidEsales\EshopCommunity\Internal\Framework\Module\Configuration\Bridge\ModuleConfigurationDaoBridge::get() must be of the type string, null given, called in /var/www/OXID/vendor/oxid-esales/oxideshop-ce/source/Core/Module/Module.php on line 108

/var/www/OXID/vendor/oxid-esales/oxideshop-ce/source/Internal/Framework/Module/Configuration/Bridge/ModuleConfigurationDaoBridge.php:50
/var/www/OXID/vendor/oxid-esales/oxideshop-ce/source/Core/Module/Module.php:108
/__w/oxid-module-internals/oxid-module-internals/Core/InternalModule.php:43
/__w/oxid-module-internals/oxid-module-internals/Controller/Admin/UtilsController.php:57
/__w/oxid-module-internals/oxid-module-internals/Controller/Admin/UtilsController.php:126
/var/www/OXID/source/modules/oxcom/moduleinternals/Tests/Integration/Controller/UtilsControllerTest.php:157
/var/www/OXID/vendor/oxid-esales/testing-library/library/UnitTestCase.php:152

5) OxidCommunity\ModuleInternals\Tests\Integration\core\FixHelperTest::testFixTemplate
OxidEsales\Eshop\Core\Exception\SystemComponentException: Function 'getClassProviderStorage' does not exist or is not accessible! (OxidCommunity\ModuleInternals\Core\ModuleStateFixer)


/var/www/OXID/vendor/oxid-esales/oxideshop-ce/source/Core/Base.php:81
/__w/oxid-module-internals/oxid-module-internals/Core/ModuleStateFixer.php:492
/__w/oxid-module-internals/oxid-module-internals/Core/ModuleStateFixer.php:499
/__w/oxid-module-internals/oxid-module-internals/Core/ModuleStateFixer.php:127
/var/www/OXID/source/modules/oxcom/moduleinternals/Tests/Integration/Core/FixHelperTest.php:81
/var/www/OXID/source/modules/oxcom/moduleinternals/Tests/Integration/Core/FixHelperTest.php:31
/var/www/OXID/vendor/oxid-esales/testing-library/library/UnitTestCase.php:152

6) OxidCommunity\ModuleInternals\Tests\Integration\core\FixHelperTest::testFixVersion
OxidEsales\Eshop\Core\Exception\SystemComponentException: Function 'getClassProviderStorage' does not exist or is not accessible! (OxidCommunity\ModuleInternals\Core\ModuleStateFixer)


/var/www/OXID/vendor/oxid-esales/oxideshop-ce/source/Core/Base.php:81
/__w/oxid-module-internals/oxid-module-internals/Core/ModuleStateFixer.php:492
/__w/oxid-module-internals/oxid-module-internals/Core/ModuleStateFixer.php:499
/__w/oxid-module-internals/oxid-module-internals/Core/ModuleStateFixer.php:127
/var/www/OXID/source/modules/oxcom/moduleinternals/Tests/Integration/Core/FixHelperTest.php:81
/var/www/OXID/source/modules/oxcom/moduleinternals/Tests/Integration/Core/FixHelperTest.php:39
/var/www/OXID/vendor/oxid-esales/testing-library/library/UnitTestCase.php:152

7) OxidCommunity\ModuleInternals\Tests\Integration\core\FixHelperTest::testFixExtensions
OxidEsales\Eshop\Core\Exception\SystemComponentException: Function 'getClassProviderStorage' does not exist or is not accessible! (OxidCommunity\ModuleInternals\Core\ModuleStateFixer)


/var/www/OXID/vendor/oxid-esales/oxideshop-ce/source/Core/Base.php:81
/__w/oxid-module-internals/oxid-module-internals/Core/ModuleStateFixer.php:492
/__w/oxid-module-internals/oxid-module-internals/Core/ModuleStateFixer.php:499
/__w/oxid-module-internals/oxid-module-internals/Core/ModuleStateFixer.php:127
/var/www/OXID/source/modules/oxcom/moduleinternals/Tests/Integration/Core/FixHelperTest.php:81
/var/www/OXID/source/modules/oxcom/moduleinternals/Tests/Integration/Core/FixHelperTest.php:48
/var/www/OXID/vendor/oxid-esales/testing-library/library/UnitTestCase.php:152

8) OxidCommunity\ModuleInternals\Tests\Integration\core\FixHelperTest::testCleanUpControllers
OxidEsales\Eshop\Core\Exception\SystemComponentException: Function 'getClassProviderStorage' does not exist or is not accessible! (OxidCommunity\ModuleInternals\Core\ModuleStateFixer)


/var/www/OXID/vendor/oxid-esales/oxideshop-ce/source/Core/Base.php:81
/__w/oxid-module-internals/oxid-module-internals/Core/ModuleStateFixer.php:492
/__w/oxid-module-internals/oxid-module-internals/Core/ModuleStateFixer.php:499
/__w/oxid-module-internals/oxid-module-internals/Core/ModuleStateFixer.php:127
/var/www/OXID/source/modules/oxcom/moduleinternals/Tests/Integration/Core/FixHelperTest.php:81
/var/www/OXID/source/modules/oxcom/moduleinternals/Tests/Integration/Core/FixHelperTest.php:57
/var/www/OXID/vendor/oxid-esales/testing-library/library/UnitTestCase.php:152

9) OxidCommunity\ModuleInternals\Tests\Integration\core\FixHelperTest::testFixControllers
OxidEsales\Eshop\Core\Exception\SystemComponentException: Function 'getClassProviderStorage' does not exist or is not accessible! (OxidCommunity\ModuleInternals\Core\ModuleStateFixer)


/var/www/OXID/vendor/oxid-esales/oxideshop-ce/source/Core/Base.php:81
/__w/oxid-module-internals/oxid-module-internals/Core/ModuleStateFixer.php:492
/__w/oxid-module-internals/oxid-module-internals/Core/ModuleStateFixer.php:499
/__w/oxid-module-internals/oxid-module-internals/Core/ModuleStateFixer.php:127
/var/www/OXID/source/modules/oxcom/moduleinternals/Tests/Integration/Core/FixHelperTest.php:81
/var/www/OXID/source/modules/oxcom/moduleinternals/Tests/Integration/Core/FixHelperTest.php:66
/var/www/OXID/vendor/oxid-esales/testing-library/library/UnitTestCase.php:152

ERRORS!
Tests: 16, Assertions: 9, Errors: 9.
##[error]Process completed with exit code 2.
```

--- a/Core/InternalModule.php
+++ b/Core/InternalModule.php
@@ -39,6 +39,7 @@ class InternalModule extends InternalModule_parent
     {
         $this->checked = false;
         $this->state = self::FINE;
+        $this->metaDataVersion = null;
         $res = parent::load($id);
         return $res;
     }

--- a/Core/InternalModule.php
+++ b/Core/InternalModule.php
@@ -37,7 +37,6 @@ class InternalModule extends InternalModule_parent
 
     public function load($id)
     {
-        $this->metaDataVersion = 0;
         $this->checked = false;
         $this->state = self::FINE;
         $res = parent::load($id);

--- a/Core/ModuleStateFixer.php
+++ b/Core/ModuleStateFixer.php
@@ -5,6 +5,7 @@ namespace OxidCommunity\ModuleInternals\Core;
 use OxidEsales\Eshop\Core\Config;
 use OxidEsales\Eshop\Core\DatabaseProvider;
 use OxidEsales\Eshop\Core\Registry;
+use OxidEsales\Eshop\Core\Routing\Module\ClassProviderStorage;
 use OxidEsales\Eshop\Core\SettingsHandler;
 use OxidEsales\Eshop\Core\Module\Module;
 use OxidEsales\Eshop\Core\Module\ModuleInstaller;
@@ -758,6 +759,74 @@ class ModuleStateFixer extends ModuleInstaller
     {
         $validator = $this->getModuleMetadataValidator();
         $validator->checkModuleExtensionsForIncorrectNamespaceClasses($module);
+    }
+
+    /**
+     * @return object
+     */
+    protected function getClassProviderStorage()
+    {
+        $classProviderStorage = oxNew(ClassProviderStorage::class);
+
+        return $classProviderStorage;
+    }
+
+    /**
+     * Remove controllers map for a given module Id from config
+     *
+     * @param string $moduleId The Id of the module
+     */
+    protected function deleteModuleControllers($moduleId)
+    {
+        $moduleControllerProvider = $this->getClassProviderStorage();
+        $moduleControllerProvider->remove($moduleId);
+    }
+
+    /**
+     * @return \OxidEsales\Eshop\Core\Contract\ControllerMapProviderInterface
+     */
+    protected function getModuleControllerMapProvider()
+    {
+        return oxNew(\OxidEsales\Eshop\Core\Routing\ModuleControllerMapProvider::class);
+    }
+    /**
+     * @return \OxidEsales\Eshop\Core\Contract\ControllerMapProviderInterface
+     */
+    protected function getShopControllerMapProvider()
+    {
+        return oxNew(\OxidEsales\Eshop\Core\Routing\ShopControllerMapProvider::class);
+    }
+
+    /**
+     * Ensure integrity of the controllerMap before storing it.
+     * Both keys and values must be unique with in the same shop or sub-shop.
+     *
+     * @param array $moduleControllers
+     *
+     * @throws ModuleValidationException
+     */
+    protected function validateModuleMetadataControllersOnActivation($moduleControllers)
+    {
+        $moduleControllerMapProvider = $this->getModuleControllerMapProvider();
+        $shopControllerMapProvider = $this->getShopControllerMapProvider();
+        $moduleControllerMap = $moduleControllerMapProvider->getControllerMap();
+        $shopControllerMap = $shopControllerMapProvider->getControllerMap();
+        $existingMaps = array_merge($moduleControllerMap, $shopControllerMap);
+        /**
+         * Ensure, that controller keys are unique.
+         * As keys are always stored in lower case, we must test against lower case keys here as well
+         */
+        $duplicatedKeys = array_intersect_key(array_change_key_case($moduleControllers, CASE_LOWER), $existingMaps);
+        if (!empty($duplicatedKeys)) {
+            throw new \OxidEsales\Eshop\Core\Exception\ModuleValidationException(implode(',', $duplicatedKeys));
+        }
+        /**
+         * Ensure, that controller values are unique.
+         */
+        $duplicatedValues = array_intersect($moduleControllers, $existingMaps);
+        if (!empty($duplicatedValues)) {
+            throw new \OxidEsales\Eshop\Core\Exception\ModuleValidationException(implode(',', $duplicatedValues));
+        }
     }
 
     /**

--- a/Tests/Integration/Controller/UtilsControllerTest.php
+++ b/Tests/Integration/Controller/UtilsControllerTest.php
@@ -67,6 +67,7 @@ class UtilsControllerTest extends UnitTestCase
     public function testGetModuleInstaller()
     {
         $utilsController = oxNew(\OxidCommunity\ModuleInternals\Controller\Admin\UtilsController::class);
+        $_POST['oxid'] = 'moduleinternals';
 
         $module = $utilsController->getModuleInstaller();
 
@@ -110,6 +111,7 @@ class UtilsControllerTest extends UnitTestCase
      */
     public function testActivateModule()
     {
+        $_POST['oxid'] = 'moduleinternals';
         $moduleInstaller = $this->getMock(
             \OxidEsales\Eshop\Core\Module\ModuleInstaller::class,
             ['activate'],
@@ -136,6 +138,7 @@ class UtilsControllerTest extends UnitTestCase
      */
     public function testDeactivateModule()
     {
+        $_POST['oxid'] = 'moduleinternals';
         $moduleInstaller = $this->getMock(
             \OxidEsales\Eshop\Core\Module\ModuleInstaller::class,
             ['deactivate'],

